include/aws/cal/der.h

```diff
@@ -0,0 +1,249 @@
+ */
+AWS_CAL_API int aws_der_encoder_write_integer(struct aws_der_encoder *encoder, struct aws_byte_cursor integer);
+/**
+ * Writes a boolean to the DER stream
+ * @param encoder The encoder to use
+ * @param boolean The boolean to write
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_write_boolean(struct aws_der_encoder *encoder, bool boolean);
+
+/**
+ * Writes a NULL token to the stream
+ * @param encoder The encoder to write to
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_write_null(struct aws_der_encoder *encoder);
+
+/**
+ * Writes a BIT_STRING to the stream
+ * @param encoder The encoder to use
+ * @param bit_string The bit string to encode
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_write_bit_string(struct aws_der_encoder *encoder, struct aws_byte_cursor bit_string);
+
+/**
+ * Writes a string to the stream
+ * @param encoder The encoder to use
+ * @param octet_string The string to encode
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_write_octet_string(
+    struct aws_der_encoder *encoder,
+    struct aws_byte_cursor octet_string);
+
+/**
+ * Begins a SEQUENCE of objects in the DER stream
+ * @param encoder The encoder to use
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_begin_sequence(struct aws_der_encoder *encoder);
+
+/**
+ * Finishes a SEQUENCE and applies it to the DER stream buffer
+ * @param encoder The encoder to update
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_end_sequence(struct aws_der_encoder *encoder);
+
+/**
+ * Begins a SET of objects in the DER stream
+ * @param encoder The encoder to use
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_begin_set(struct aws_der_encoder *encoder);
+
+/**
+ * Finishes a SET and applies it to the DER stream buffer
+ * @param encoder The encoder to update
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_end_set(struct aws_der_encoder *encoder);
+
+/**
+ * Retrieves the contents of the encoder stream buffer
+ * @param encoder The encoder to read from
+ * @param cursor The cursor to point at the stream buffer
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_encoder_get_contents(struct aws_der_encoder *encoder, struct aws_byte_cursor *contents);
+
+/**
+ * Initializes an DER decoder
+ * @param decoder The decoder to initialize
+ * @param allocator The allocator to use
+ * @param buffer The DER formatted buffer to parse
+ * @return AWS_OP_ERR if an error occurs, otherwise AWS_OP_SUCCESS
+ */
+AWS_CAL_API int aws_der_decoder_init(
```

Comment:
Proposal for alternative decoder API:
- Single function: `int aws_der_decode(buffer, allocator?, vtable, userdata)`
- Function decodes entire buffer, invokes appropriate vtable functions as it goes. Ex: on_int, on_string
- Vtable functions can halt decoder via return value if something goes wrong
- Or instead of a vtable there's one callback fn like: `bool aws_der_decode_fn(aws_der_type, int, cursor, bool, userdata)` and the user checks the int/cursor/bool based on the der-type.

Advantages of this proposal:
- User can't mis-use API. Can't forget to call parse(). Can't accidentally call get_string() when the next element isn't a string.
- Might avoid need for allocations? Since we're invoking callbacks as we go? Rather than parsing all at once and saving it for later?

Reply:
Can totally do that, @JonathanHenson thoughts?

Reply:
I don't know what we'd need a v-table for any of this. 

I don't mind unifying init and parse though. I don't think we need a streaming parser.